### PR TITLE
Yoyo exam submission (branch name: Initial)

### PR DESCRIPTION
**Pull request by syvlabs**
Hi @zakame, good day! Here is my submission for the YOYO technical exam.
If you would like to check out the release APK, you can download it [here](http://syvlabs.com/static/organizr-1.0.apk). I would also like to discuss some details about my app:

**Focus**
Since this is a note-taking and todo app, I created this app specifically with usability in mind. Besides being functional, the user also has to _like_ using the app. Hence I went for an app which feels quick, easy to use, and is visually pleasing. (The visual design that I made mostly conforms to Google's material design guidelines, and all the icons were from Google as well.) In consequence, I also had to focus on transitions and animations in order to make it feel quick, which added some complexity to the app. As I explain the app, I will mention how my decisions relate to this focus.

**Design Patterns**
This app is an interaction-heavy application (as opposed to data-heavy), hence there are only a few cases where I have to instantiate objects. Database transactions were carried out through a factory and a helper class, in order to separate the view from data-heavy operations. The interactions and UI part conforms to Android's MVC patterns (Adapter, ViewHolder, etc).
One of the design decisions I made was to select whether to use an Activity or a Fragment. The current app uses a single main activity that shows the Notes and Todos. When editing entries however, it launches a fragment in the form of a dialogfragment. I specifically chose a dialog instead of a new activity, since I wanted to give the user the impression that when editing a note/to-do, he/she can immediately return to the main activity. This is because a dialog lays the UI on top of the main activity, while if you use a new activity it will replace the entire screen. The dialog also gives the impression of a parent-child relationship between the main activity and the dialog.

**Libraries**
I was liberal in the use of several libraries, as long as it did not go over the size limit.
Here are the libraries I used:
- Support libraries: com.android.support.(appcompat, design, recyclerview, support, cardview)
  - Since I am supporting until KitKat and I want to follow material design, there was no choice but to include these libraries. Some of the UI elements I included were: recyclerview, cardview, viewpager, etc.
- Butterknife: com.jakewharton:butterknife
  - For easy and clean view bindings, as well as listener bindings. There are some more similar libraries out there, but in my opinion this is the cleanest one. This makes views much cleaner and more readable as compared to findViewById.
- Greendao: de.greenrobot:greendao(-generator)
  - There are several persistent storage libraries available, and I haven't tested them all, but in my experience, Greendao suffices for most cases. A note-taking app such as this one does not have very fast read/write requirements, hence I chose the most straightforward one. I also believe that Greendao has more documentation and support compared to the other libs.
- Rebound: com.facebook.rebound
  - Instead of resorting to the usual android animations, I used Facebook's rebound library for many of my transitions. In my experience, the Facebook messenger app is one of the most quick-looking apps that I have ever used, due to its "springy" animations. I wanted the same zippy user experience in this app.
- Otto: com.squareup:otto
  - Otto is an event bus for android. There are alternatives such as Rx, but since my use case is simple, I opted for Otto, since it is much cleaner than Rx. This event bus was used to automatically update the notes and to-dos on database changes.
- Spectrum: com.thebluealliance:spectrum:0.5.0
  - This is just a color selection library for android, which drives the palette/color-picker for the notes and to-dos. There might be better libraries around, but I settled with the first one I found that works and is simple enough, since the color-picker is a minor part of the application.

**Challenges**
Most challenges I had were on the UI/UX side, specifically on how the app should be designed in order to make the app usable and intuitive. For instance, where should the delete buttons be placed? or, how should the user edit the entries? These kinds of UX problems took a lot of time to implement and try out. Some other problems were on backwards compatibility (with KitKat) and some Android bugs (control of soft keyboard).

**Additional Features**
If I had more time, I would like to introduce syncing in order to make it a fully-fledged app, but that would take much more time. It would also be interesting to add notifications to the to-do app.

**Copyright, Legal, etc**
I implemented as well as designed the whole app, and the libraries I used are all free to use. All graphics and icons are from [Google Design](https://design.google.com/icons/).

### CODE REVIEW

**Comment by zakame** [comment]
Thanks @syvlabs !  We'll set an review/interview soon :muscle:

File: gradlew [hunk 1]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
```
Comment (zakame) [inline gradlew:1]: This file should be set executable in OS X / Linux.

Comment (syvlabs) [inline gradlew:1]: Sorry about that, I was working on a windows machine and didn't notice
that. Should I try to fix it?

Comment (zakame) [inline gradlew:1]: No worries :+1:

**Comment by zakame** [comment]
LGTM

**Pull request closed without merging**